internal/services/v1/lookupwatch.go

```diff
@@ -0,0 +1,303 @@
+				usagemetrics.UnaryServerInterceptor(),
+			),
+			Stream: grpcmw.ChainStreamServer(
+				grpcvalidate.StreamServerInterceptor(),
+				handwrittenvalidation.StreamServerInterceptor,
+				usagemetrics.StreamServerInterceptor(),
+			),
+		},
+	}
+}
+
+type lookupWatchServer struct {
+	v1lookupwatch.UnimplementedLookupWatchServiceServer
+	shared.WithServiceSpecificInterceptors
+
+	dispatch     dispatchpkg.Dispatcher
+	defaultDepth uint32
+}
+
+func (lw *lookupWatchServer) WatchAccessibleResources(req *v1lookupwatch.WatchAccessibleResourcesRequest, stream v1lookupwatch.LookupWatchService_WatchAccessibleResourcesServer) error {
+	ctx := stream.Context()
+
+	ds := datastoremw.MustFromContext(ctx)
+	objectTypesMap := make(map[string]struct{})
+	for _, objectType := range []string{req.ResourceObjectType} {
+		objectTypesMap[objectType] = struct{}{}
+	}
+
+	var afterRevision decimal.Decimal
+	if req.OptionalStartTimestamp != nil && req.OptionalStartTimestamp.Token != "" {
+		decodedRevision, err := zedtoken.DecodeRevision(req.OptionalStartTimestamp)
+		if err != nil {
+			return status.Errorf(codes.InvalidArgument, "failed to decode start revision: %s", err)
+		}
+
+		afterRevision = decodedRevision
+	} else {
+		var err error
+		afterRevision, err = ds.HeadRevision(ctx)
+		if err != nil {
+			return status.Errorf(codes.Unavailable, "failed to start watch: %s", err)
+		}
+	}
+
+	usagemetrics.SetInContext(ctx, &dispatchv1.ResponseMeta{
+		DispatchCount: 1,
+	})
+
+	updates, errchan := ds.Watch(ctx, afterRevision)
+
+	for {
+		select {
+		case update, ok := <-updates:
+			if ok {
+				return lw.processWatchResponse(&ctx, req, update, &stream)
+			}
+		case err := <-errchan:
+			switch {
+			case errors.As(err, &datastore.ErrWatchCanceled{}):
+				return status.Errorf(codes.Canceled, "watch canceled by user: %s", err)
+			case errors.As(err, &datastore.ErrWatchDisconnected{}):
+				return status.Errorf(codes.ResourceExhausted, "watch disconnected: %s", err)
+			default:
+				return status.Errorf(codes.Internal, "watch error: %s", err)
+			}
+		}
+	}
+}
+
+// Called at each notification of the Watch API
+func (lw *lookupWatchServer) processWatchResponse(
+	ctx *context.Context,
+	req *v1lookupwatch.WatchAccessibleResourcesRequest,
+	updates *datastore.RevisionChanges,
+	stream *v1lookupwatch.LookupWatchService_WatchAccessibleResourcesServer,
+) error {
+	for _, update := range updates.Changes {
+		if err := lw.processUpdate(ctx, req, update, &updates.Revision, stream); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Called every time a relation is changed in spicedb
+func (lw *lookupWatchServer) processUpdate(
+	ctx *context.Context,
+	req *v1lookupwatch.WatchAccessibleResourcesRequest,
+	update *core.RelationTupleUpdate,
+	atRevision *datastore.Revision,
+	stream *v1lookupwatch.LookupWatchService_WatchAccessibleResourcesServer,
+) error {
+
+	ds := datastoremw.MustFromContext((*stream).Context())
+
+	// STEP 1: CALL LOOKUPSUBJECTS
```

Comment:
This is a long function, so can you expand these step comments to explain what you are doing, and why?